File: engine.py
import os, shutil, threading
from pathlib import Path
import chess
import chess.engine

class EngineManager:
    def __init__(self):
        self.stockfish_path = self.find_stockfish()
        self.engine = None
        self.lock = threading.RLock()
        self.depth = 24

    def find_stockfish(self):
        candidates=[]
        env=os.getenv("STOCKFISH_PATH")
        if env: candidates.append(env)
        names=("stockfish-windows-x86-64-universal.exe","stockfish-windows-x86-64-avx2.exe","stockfish.exe","stockfish")
        for name in names:
            found=shutil.which(name)
            if found: candidates.append(found)
        root=Path(__file__).resolve().parent
        for name in names:
            p=root/name
            if p.exists(): candidates.append(str(p))
        for p in root.rglob("stockfish*.exe"): candidates.append(str(p))
        for p in root.rglob("stockfish"):
            if p.is_file(): candidates.append(str(p))
        for p in candidates:
            try:
                if Path(p).is_file(): return str(Path(p).resolve())
            except OSError: pass
        return None

    def open(self):
        if not self.stockfish_path: return None
        with self.lock:
            if self.engine is None:
                self.engine=chess.engine.SimpleEngine.popen_uci(self.stockfish_path)
                threads=max(1,min(8,(os.cpu_count() or 4)-1))
                try: self.engine.configure({"Threads":threads,"Hash":512})
                except Exception: pass
            return self.engine

    def close(self):
        with self.lock:
            if self.engine:
                try: self.engine.quit()
                except Exception: pass
                self.engine=None

    def choose_move(self, board, bot):
        if bot=="Beginner":
            import random
            return random.choice(list(board.legal_moves)) if board.legal_moves else None
        if bot=="Minimax": return self.fallback(board,3)
        eng=self.open()
        if eng:
            limit=chess.engine.Limit(time=1.2 if bot=="Stockfish" else 4.0, depth=24 if bot=="Stockfish" else 32)
            opts={"Skill Level":20}
            if bot=="Magnus": opts.update({"Threads":max(1,min(8,(os.cpu_count() or 4)-1)),"Hash":512})
            try:
                with self.lock: return eng.play(board,limit,options=opts).move
            except Exception: pass
        return self.fallback(board,4 if bot=="Stockfish" else 5)

    def analyze(self, board, bot="Stockfish", time_limit=0.8, depth=24, multipv=10):
        eng=self.open()
        if eng:
            try:
                with self.lock:
                    infos=eng.analyse(board,chess.engine.Limit(time=time_limit,depth=depth),multipv=multipv)
                out=[]
                for info in infos:
                    score=info["score"].pov(board.turn); cp=score.score(mate_score=100000); pv=info.get("pv",[])
                    out.append({"move":pv[0] if pv else None,"cp":cp,"mate":score.mate(),"depth":info.get("depth",depth),"pv":pv[:8]})
                return out
            except Exception: pass
        m=self.fallback(board,4)
        return [{"move":m,"cp":0,"mate":None,"depth":4,"pv":[m] if m else []}]

    def analyze_move(self,before,played,bot):
        # MultiPV 10 is used for ranking; the best move is independently searched so MultiPV does not weaken it.
        top=self.analyze(before,"Stockfish",1.15,26,10)
        best=top[0] if top else {"move":played,"cp":0,"mate":None,"depth":0,"pv":[played]}
        eng=self.open(); played_cp=best.get("cp",0)
        after=before.copy(); after.push(played)
        if eng:
            try:
                with self.lock:
                    info=eng.analyse(after,chess.engine.Limit(time=1.0,depth=24))
                opp=info["score"].pov(after.turn); played_cp=-opp.score(mate_score=100000)
            except Exception: pass
        best_cp=best.get("cp",0); loss=max(0,best_cp-played_cp)
        label=self.classify(before,played,best.get("move"),loss,top)
        rank=next((i+1 for i,x in enumerate(top) if x.get("move")==played),None)
        return {"move":played,"best":best.get("move"),"best_cp":best_cp,"played_cp":played_cp,"loss":loss,"label":label,"top":top,"depth":best.get("depth",0),"rank":rank,"before_fen":before.fen(),"after_fen":after.fen()}

    def classify(self, board, move, best, loss, top):
        # Keep BEST separate from GREAT: matching the engine's top move is BEST,
        # while GREAT is reserved for a very close alternative that is not the
        # engine's exact first choice. This prevents every accurate move from
        # being shown as GREAT.
        if self.is_book(board, move):
            return "BOOK MOVE"

        rank = next((i + 1 for i, item in enumerate(top) if item.get("move") == move), 99)

        # Brilliant is intentionally rare: a non-best move must contain a
        # concrete tactical idea and lose almost nothing in the engine score.
        if move != best and board.is_capture(move) and board.gives_check(move) and loss <= 18:
            return "BRILLIANT MOVE"

        if move == best:
            return "BEST MOVE"

        if rank <= 2 and loss <= 10:
            return "GREAT MOVE"
        if rank <= 3 and loss <= 25:
            return "EXCELLENT MOVE"
        if rank <= 5 and loss <= 50:
            return "GOOD MOVE"
        if loss <= 100:
            return "INACCURACY"
        if loss <= 250:
            return "MISTAKE"
        return "BLUNDER"

    def is_book(self, board, move):
        # Small position-aware opening book for common first moves. It only marks
        # a move as BOOK when the preceding sequence matches a known opening line.
        lines = {
            (): {"e2e4", "d2d4", "c2c4", "g1f3"},
            ("e2e4",): {"e7e5", "c7c5", "e7e6", "c7c6", "g8f6"},
            ("e2e4", "e7e5"): {"g1f3", "f1c4", "d2d4"},
            ("e2e4", "e7e5", "g1f3"): {"b8c6", "g8f6", "d7d6"},
            ("e2e4", "e7e5", "g1f3", "b8c6"): {"f1b5", "f1c4", "d2d4"},
            ("e2e4", "c7c5"): {"g1f3", "b1c3", "c2c3"},
            ("e2e4", "c7c5", "g1f3"): {"d7d6", "b8c6", "e7e6"},
            ("d2d4",): {"d7d5", "g8f6", "g7g6", "e7e6"},
            ("d2d4", "d7d5"): {"c2c4", "g1f3", "c1f4", "c2c3"},
            ("d2d4", "g8f6"): {"c2c4", "c1f4", "g1f3"},
            ("c2c4",): {"e7e5", "g8f6", "e7e6", "c7c5"},
            ("g1f3",): {"d7d5", "g8f6", "c7c5", "e7e6"},
        }
        key = tuple(m.uci() for m in list(board.move_stack))
        allowed = lines.get(key)
        return move.uci() in allowed if allowed else False

    def fallback(self,board,depth):
        vals={chess.PAWN:100,chess.KNIGHT:320,chess.BISHOP:330,chess.ROOK:500,chess.QUEEN:900,chess.KING:0}
        def ev(b): return sum((1 if p.color else -1)*vals[p.piece_type] for p in b.piece_map().values())
        def mm(b,d,a,beta,maxi):
            if d==0 or b.is_game_over(): return ev(b),None
            best=(-10**9,None) if maxi else (10**9,None)
            for m in b.legal_moves:
                b.push(m); sc,_=mm(b,d-1,a,beta,not maxi); b.pop()
                if (maxi and sc>best[0]) or ((not maxi) and sc<best[0]): best=(sc,m)
                if maxi: a=max(a,sc)
                else: beta=min(beta,sc)
                if beta<=a: break
            return best
        return mm(board,depth,-10**9,10**9,board.turn==chess.WHITE)[1]

File: engine.js
import { Chess } from 'chess.js'

const VALUE = { p: 100, n: 320, b: 330, r: 500, q: 900, k: 20000 }
const PST = {
  p: [0,0,0,0,0,0,0,0, 50,50,50,50,50,50,50,50, 10,10,20,30,30,20,10,10, 5,5,10,25,25,10,5,5, 0,0,0,20,20,0,0,0, 5,-5,-10,0,0,-10,-5,5, 5,10,10,-20,-20,10,10,5, 0,0,0,0,0,0,0,0],
  n: [-50,-40,-30,-30,-30,-30,-40,-50, -40,-20,0,0,0,0,-20,-40, -30,0,10,15,15,10,0,-30, -30,5,15,20,20,15,5,-30, -30,0,15,20,20,15,0,-30, -30,5,10,15,15,10,5,-30, -40,-20,0,5,5,0,-20,-40, -50,-40,-30,-30,-30,-30,-40,-50],
  b: [-20,-10,-10,-10,-10,-10,-10,-20, -10,0,0,0,0,0,0,-10, -10,0,5,10,10,5,0,-10, -10,5,5,10,10,5,5,-10, -10,0,10,10,10,10,0,-10, -10,10,10,10,10,10,10,-10, -10,5,0,0,0,0,5,-10, -20,-10,-10,-10,-10,-10,-10,-20],
  r: [0,0,0,5,5,0,0,0, -5,0,0,0,0,0,0,-5, -5,0,0,0,0,0,0,-5, -5,0,0,0,0,0,0,-5, -5,0,0,0,0,0,0,-5, 5,10,10,10,10,10,10,5, 0,0,0,0,0,0,0,0],
  q: [-20,-10,-10,-5,-5,-10,-10,-20, -10,0,0,0,0,0,0,-10, -10,0,5,5,5,5,0,-10, -5,0,5,5,5,5,0,-5, 0,0,5,5,5,5,0,-5, -10,5,5,5,5,5,0,-10, -20,-10,-10,-5,-5,-10,-10,-20],
  k: [-30,-40,-40,-50,-50,-40,-40,-30, -30,-40,-40,-50,-50,-40,-40,-30, -30,-40,-40,-50,-50,-40,-40,-30, -30,-40,-40,-50,-50,-40,-40,-30, -20,-30,-30,-40,-40,-30,-30,-20, -10,-20,-20,-20,-20,-20,-20,-10, 20,20,0,0,0,0,20,20]
}

function idx(file, rank){ return (7-rank)*8 + file }
function pieceScore(piece, file, rank){
  const table = PST[piece.type]
  const square = piece.color === 'w' ? idx(file, rank) : idx(file, 7-rank)
  return VALUE[piece.type] + (table ? table[square] : 0)
}

export function evaluate(chess){
  let score = 0
  for(let r=0;r<8;r++) for(let f=0;f<8;f++){
    const sq = String.fromCharCode(97+f)+(r+1)
    const p = chess.get(sq)
    if(p) score += (p.color==='w'?1:-1)*pieceScore(p,f,r)
  }
  if(chess.isCheck()) score += chess.turn()==='w' ? -45 : 45
  if(chess.isCheckmate()) score += chess.turn()==='w' ? -30000 : 30000
  return score
}

function orderedMoves(chess){
  return chess.moves({ verbose:true }).sort((a,b)=>{
    const av=(a.captured?VALUE[a.captured]:0)+(a.promotion?VALUE[a.promotion]:0)+(a.san.includes('+')?80:0)
    const bv=(b.captured?VALUE[b.captured]:0)+(b.promotion?VALUE[b.promotion]:0)+(b.san.includes('+')?80:0)
    return bv-av
  })
}

function search(chess, depth, alpha, beta){
  if(depth===0 || chess.isGameOver()) return {score:evaluate(chess)}
  const moves=orderedMoves(chess)
  if(!moves.length) return {score:evaluate(chess)}
  const maximizing=chess.turn()==='w'
  let bestScore=maximizing?-Infinity:Infinity
  let bestMove=moves[0]
  for(const move of moves){
    chess.move(move)
    const result=search(chess,depth-1,alpha,beta)
    chess.undo()
    if(maximizing){
      if(result.score>bestScore){bestScore=result.score;bestMove=move}
      alpha=Math.max(alpha,bestScore)
    }else{
      if(result.score<bestScore){bestScore=result.score;bestMove=move}
      beta=Math.min(beta,bestScore)
    }
    if(beta<=alpha) break
  }
  return {score:bestScore,move:bestMove}
}

export function getBestMove(fen, depth=2){
  const chess=new Chess(fen)
  const result=search(chess,Math.max(1,Math.min(4,depth)),-Infinity,Infinity)
  return result.move ? {uci: result.move.from+result.move.to+(result.move.promotion||''), san:result.move.san, score:result.score} : null
}

export function rankMove(fen, uci, depth=2){
  const chess=new Chess(fen)
  const moves=orderedMoves(chess)
  const ranked=[]
  for(const move of moves){
    chess.move(move)
    const result=search(chess,Math.max(0,depth-1),-Infinity,Infinity).score
    chess.undo()
    const scoreForMover=chess.turn()==='w'?result:-result
    ranked.push({uci:move.from+move.to+(move.promotion||''),score:scoreForMover,san:move.san})
  }
  ranked.sort((a,b)=>b.score-a.score)
  const idx=ranked.findIndex(m=>m.uci===uci)
  const best=ranked[0]
  const played=idx>=0?ranked[idx]:{score:best?.score||0}
  return {rank:idx<0?99:idx+1,best,bestScore:best?.score||0,playedScore:played.score,loss:Math.max(0,(best?.score||0)-played.score)}
}

export function classifyMove(fen, uci, san, depth=2){
  const chess=new Chess(fen)
  const best=getBestMove(fen,depth)
  const stats=rankMove(fen,uci,depth)
  if(chess.history().length<10) return {label:'BOOK MOVE',...stats,bestMove:best?.uci||''}
  if(!best) return {label:'GOOD MOVE',...stats,bestMove:''}
  if(uci===best.uci) return {label:'BEST MOVE',...stats,bestMove:best.uci}
  const move=chess.move({from:uci.slice(0,2),to:uci.slice(2,4),promotion:uci.slice(4)||undefined})
  const brilliant=move?.captured && move?.san?.includes('+') && stats.loss<=35
  if(brilliant) return {label:'BRILLIANT MOVE',...stats,bestMove:best.uci}
  if(stats.rank<=2 && stats.loss<=10) return {label:'GREAT MOVE',...stats,bestMove:best.uci}
  if(stats.rank<=3 && stats.loss<=25) return {label:'EXCELLENT MOVE',...stats,bestMove:best.uci}
  if(stats.rank<=5 && stats.loss<=50) return {label:'GOOD MOVE',...stats,bestMove:best.uci}
  if(stats.loss<=100) return {label:'INACCURACY',...stats,bestMove:best.uci}
  if(stats.loss<=250) return {label:'MISTAKE',...stats,bestMove:best.uci}
  return {label:'BLUNDER',...stats,bestMove:best.uci}
}

export function makeBotMove(fen, level){
  const depth = level==='beginner'?1:level==='intermediate'?2:3
  const best=getBestMove(fen,depth)
  if(!best) return null
  return best
}
